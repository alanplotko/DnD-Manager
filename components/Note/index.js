import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import { StyleSheet, View, Text } from 'react-native';
import { COLOR, Icon, IconToggle } from 'react-native-material-ui';
import DefaultTheme from 'FifthEditionManager/themes/DefaultTheme';

const Note = (props) => {
  // Note style and icon setup
  let iconColor = COLOR.black; // Default
  let noteTypeStyle = null;
  if (props.type === 'error') {
    iconColor = COLOR.red500;
    noteTypeStyle = styles.error;
  } else if (props.type === 'tip') {
    iconColor = COLOR.blueGrey500;
    noteTypeStyle = styles.tip;
  } else if (props.type === 'info') {
    iconColor = COLOR.lightBlue500;
    noteTypeStyle = styles.info;
  }

  // Theme setup
  const { textColor } = props.uiTheme.palette;
  const textStyle = { color: textColor };

  return (
    <View style={[styles.note, noteTypeStyle]}>
      {
        props.collapsible &&
        <View style={{ position: 'absolute', top: 0, right: 0 }}>
          <IconToggle
            name={props.isCollapsed ? 'arrow-drop-down' : 'arrow-drop-up'}
            color={iconColor}
            size={28}
            percent={50}
            onPress={() => props.toggleNoteHandler()}
          />
        </View>
      }
      <View
        style={{
          flex: 1,
          flexDirection: 'row',
          marginBottom: !props.isCollapsed || !props.collapsible ? 10 : 0,
        }}
      >
        {
          props.icon &&
          <Icon
            name={props.icon}
            color={iconColor}
            style={{ marginRight: 10 }}
          />
        }
        <Text style={[styles.noteHeading, textStyle]}>
          {props.title}
        </Text>
      </View>
      {
        (!props.collapsible || !props.isCollapsed) &&
        <View>
          <Text style={[styles.noteText, textStyle]}>
            {props.children}
          </Text>
          {
            props.type === 'error' &&
            <Text style={styles.errorTimestamp}>
              Last attempted {moment().format('h:mm:ss A')}
            </Text>
          }
        </View>
      }
    </View>
  );
};

const styles = StyleSheet.create({
  note: {
    borderWidth: 0.5,
    borderRadius: 3,
    marginBottom: 15,
    padding: 15,
  },
  noteHeading: {
    fontFamily: 'RobotoBold',
    color: COLOR.black,
    fontSize: 18,
  },
  noteText: {
    fontFamily: 'Roboto',
    color: COLOR.black,
    fontSize: 14,
  },
  errorTimestamp: {
    textAlign: 'right',
    fontStyle: 'italic',
    marginTop: 10,
  },
  tip: {
    backgroundColor: COLOR.blueGrey50,
    borderColor: COLOR.blueGrey500,
  },
  error: {
    backgroundColor: COLOR.red50,
    borderColor: COLOR.red500,
  },
  info: {
    backgroundColor: COLOR.lightBlue50,
    borderColor: COLOR.lightBlue500,
  },
});

Note.propTypes = {
  title: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  icon: PropTypes.string,
  type: PropTypes.string.isRequired,
  collapsible: PropTypes.bool,
  isCollapsed: PropTypes.bool,
  // Require function only when collapsible prop is true
  toggleNoteHandler: (props, propName, componentName) => {
    const value = props[propName];
    if (props.collapsible) {
      if (!value) {
        return new Error(`Prop \`${propName}\` not supplied to \`${componentName}\`. Required when note is collapsible. Validation failed.`);
      } else if (typeof value !== 'function') {
        return new Error(`Invalid prop \`${propName}\` supplied to \`${componentName}\`. Type \`function\` expected. Validation failed.`);
      }
    }
    return null;
  },
  uiTheme: PropTypes.object,
};

Note.defaultProps = {
  icon: null,
  collapsible: false,
  isCollapsed: false,
  toggleNoteHandler: undefined,
  uiTheme: DefaultTheme,
};

export default Note;
