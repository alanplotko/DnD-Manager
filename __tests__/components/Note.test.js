import React from 'react';
import { Text } from 'react-native';
import Note from 'FifthEditionManager/components/Note';

class NoteWrapper extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: 'collapsed',
      children: <Text>expanded</Text>,
      icon: 'info',
      type: 'info',
      collapsible: true,
      isCollapsed: true,
      toggleNoteHandler: this.toggleNote.bind(this),
    };
  }

  toggleNote = () => this.setState({ isCollapsed: !this.state.isCollapsed });

  render() {
    return <Note {...this.state} />;
  }
}

describe('Note', () => {
  let timeStub;

  beforeEach(() => {
    timeStub = sinon.stub(Date, 'now').returns(0);
  });

  afterEach(() => {
    timeStub.restore();
  });

  test('renders as error note with error type', () => {
    const wrapper = shallow(<Note title="Error" icon="error" type="error"><Text>An error occurred!</Text></Note>);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders as info note with info type', () => {
    const wrapper = shallow(<Note title="Info" icon="info" type="info"><Text>Information here.</Text></Note>);
    expect(wrapper).toMatchSnapshot();
  });

  test('can expand and collapse properly', () => {
    const wrapper = shallow(<NoteWrapper />);

    // Collapsed by default
    expect(wrapper.dive()).toMatchSnapshot();
    expect(wrapper.dive().find('Text').last().dive().text()).toEqual('collapsed');

    // Toggle to expand
    wrapper.dive().find('IconToggle').props().onPress();
    wrapper.update();

    // Expanded for the first time
    expect(wrapper.dive()).toMatchSnapshot();
    expect(wrapper.dive().find('Text').last().dive().text()).toEqual('expanded');

    // Toggle to collapse
    wrapper.dive().find('IconToggle').props().onPress();
    wrapper.update();

    // Collapsed again
    expect(wrapper.dive()).toMatchSnapshot();
    expect(wrapper.dive().find('Text').last().dive().text()).toEqual('collapsed');
  });
});
