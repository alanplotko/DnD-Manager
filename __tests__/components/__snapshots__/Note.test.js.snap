// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note can expand and collapse properly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#e1f5fe",
      "borderColor": "#03a9f4",
      "borderRadius": 3,
      "borderWidth": 0.5,
      "marginBottom": 15,
      "padding": 15,
    }
  }
>
  <View
    style={
      Object {
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <IconToggle
      color="#03a9f4"
      name="arrow-drop-down"
      onPress={[Function]}
      percent={50}
      size={28}
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "marginBottom": 0,
      }
    }
  >
    <Icon
      color="#03a9f4"
      name="info"
      style={
        Object {
          "marginRight": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "RobotoBold",
            "fontSize": 18,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      collapsed
    </Text>
  </View>
</View>
`;

exports[`Note can expand and collapse properly 2`] = `
<View
  style={
    Object {
      "backgroundColor": "#e1f5fe",
      "borderColor": "#03a9f4",
      "borderRadius": 3,
      "borderWidth": 0.5,
      "marginBottom": 15,
      "padding": 15,
    }
  }
>
  <View
    style={
      Object {
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <IconToggle
      color="#03a9f4"
      name="arrow-drop-up"
      onPress={[Function]}
      percent={50}
      size={28}
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Icon
      color="#03a9f4"
      name="info"
      style={
        Object {
          "marginRight": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "RobotoBold",
            "fontSize": 18,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      collapsed
    </Text>
  </View>
  <View
    style={Object {}}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 14,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        expanded
      </Text>
    </Text>
  </View>
</View>
`;

exports[`Note can expand and collapse properly 3`] = `
<View
  style={
    Object {
      "backgroundColor": "#e1f5fe",
      "borderColor": "#03a9f4",
      "borderRadius": 3,
      "borderWidth": 0.5,
      "marginBottom": 15,
      "padding": 15,
    }
  }
>
  <View
    style={
      Object {
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <IconToggle
      color="#03a9f4"
      name="arrow-drop-down"
      onPress={[Function]}
      percent={50}
      size={28}
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "marginBottom": 0,
      }
    }
  >
    <Icon
      color="#03a9f4"
      name="info"
      style={
        Object {
          "marginRight": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "RobotoBold",
            "fontSize": 18,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      collapsed
    </Text>
  </View>
</View>
`;

exports[`Note renders as error note with error type 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffebee",
      "borderColor": "#f44336",
      "borderRadius": 3,
      "borderWidth": 0.5,
      "marginBottom": 15,
      "padding": 15,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Icon
      color="#f44336"
      name="error"
      style={
        Object {
          "marginRight": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "RobotoBold",
            "fontSize": 18,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      Error
    </Text>
  </View>
  <View
    style={Object {}}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 14,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        An error occurred!
      </Text>
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontStyle": "italic",
          "marginTop": 10,
          "textAlign": "right",
        }
      }
    >
      Last attempted 
      12:00:00 AM
    </Text>
  </View>
</View>
`;

exports[`Note renders as info note with info type 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#e1f5fe",
      "borderColor": "#03a9f4",
      "borderRadius": 3,
      "borderWidth": 0.5,
      "marginBottom": 15,
      "padding": 15,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <Icon
      color="#03a9f4"
      name="info"
      style={
        Object {
          "marginRight": 10,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "RobotoBold",
            "fontSize": 18,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      Info
    </Text>
  </View>
  <View
    style={Object {}}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Roboto",
            "fontSize": 14,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Information here.
      </Text>
    </Text>
  </View>
</View>
`;
